# Automating all this! 

## American University COMM 260, fall 2025

## 👀 What am I looking at?

This is a Jupyter Notebook. It lets you run Python code cell-by-cell. Any cell that has a bracket to the left of it is Python code you can run. The rest are written in markdown.

Go ahead and try runninng these next few cells one at a time. (Note: you can run a cell by selecting it then clicking the play button, or by hitting shfit+enter)

In [28]:
2+3

5

In [29]:
print("hello world")

hello world


In [30]:
my_name = "**Write your name here**"
print("Nice to meet you, " + my_name)

Nice to meet you, **Write your name here**


🧑‍💻 See on that cell above how it's printing `Nice to meet you, **Write your name here**`? Fix that by deleting `**Write your name here**` and putting your name instead in the next cell, then running it.

In [31]:
my_name = "**Write your name here**"
print("Nice to meet you, " + my_name)

Nice to meet you, **Write your name here**


This is using a *variable* in your code, in this case called `my_name`. That variable can be set to whatever you want. It's like algebra, where `x+3` changes depending on what `x` is.

In [32]:
x = 5

In [33]:
x+3

8

🧑‍💻 Now change the variable of x to something else and run the next 2 cells.

In [34]:
x = 

In [35]:
x+3

13

We're not going to spend a ton of time going over all the fundamentals of programming, but we'll be taking advantage of variables as "shortcuts" throughout our lesson. So, for example, if I want to keep using my name in code, I don't have to write `"Sean McMinn"` every time, I can just type `my_name`

## 📚 Loading in libraries

We can use this same idea of shorcuts for large blocks of prewritten code. These are known as *modules* or *libraries*. Some come included with Python, others you have to install.

First, we're going to bring in a module called `date` from a larger module called `datetime`

In [9]:
from datetime import date

See if you can figure out what the next block of code is doing, then run it.

In [10]:
start_date = date(2025, 1, 20)
end_date = date.today()

difference = end_date - start_date
print("It's been", difference.days, "days since Trump took office.")

It's been 288 days since Trump took office.


This module `date` lets us work with dates and store them as variables. In this case, it's also calculating the difference in dates for us.

Another library called `pandas` lets us work with spreadsheets and store them as variables. We typically abbreviate it as `pd`, which you can tell Python to do when you import the library. 

This one doesn't come standard with Python, so first we want to make sure it's installed on your computer. (We're also installing another library called `lxml` that lets pandas grab tables from the web.)

In [11]:
%pip install pandas lxml

Note: you may need to restart the kernel to use updated packages.


Now we're going to import pandas and "nickname" it `pd`. This is a common convention just so we don't have to spell out `pandas` every line of code we use it.

In [12]:
import pandas as pd

## 🧮 Dataframes

First, let's just try reading in a CSV of AU mens basketball points leaders. Make sure the csv file `points_leaders.csv` is in the same folder as this Jupyter notebook file.

In [13]:
pd.read_csv("points_leaders.csv")

Name  GP  Pts  Pts/G
0       #15 Rogers, Matt  35  586   16.7
1        #23 Jones, Greg  33  340   10.3
2   #01 Stephens, Elijah  31  330   10.6
3      #03 Smalls, Colin  33  322    9.8
4       #33 Mayock, Matt  35  254    7.3
5     #12 Sprouse, Geoff  20  138    6.9
6      #04 Ball, Lincoln  35  137    3.9
7    #07 Nausadis, Wyatt  35  107    3.1
8     #25 Michaels, Eric  29   94    3.2
9      #32 Iturbe, Julen  26   70    2.7
10  #45 Whittaker, Mason  24   11    0.5
11   #10 Samuels, Adrian  15    8    0.5
12     #05 Daniels, Alex   3    4    1.3
13       #20 Holod, Adam   3    3    1.0
14       #21 Jones, Noah   7    2    0.3

This output a *dataframe* — that's what Pandas calls a table or spreadsheet. You haven't done anything with it yet, you're just viewing it in your notebook.

See how this already has Pts/G calculated? Let's see if we can match that math.

First, store the dataframe as a variable. We'll call it `df` - short for dataframe.

In [14]:
df = pd.read_csv("points_leaders.csv")

Then, add a new column calculating our own points per game. To add a new column, the syntax is `df["whatever_you_want_to_call_your_column"] = `

We can also do math between columns using pandas. `df['Pts']/df['GP']` divides the points column by the games played column.

In [15]:
df['pts_per_game'] = df['Pts']/df['GP']

Notice nothing "happened" when you ran the two cells above. That's because you're not outputting anything at the end of the cell. You're only assigning something to a variable or a column. 

In this next cell, let's output `df` so we can look at it.

In [16]:
df

Name  GP  Pts  Pts/G  pts_per_game
0       #15 Rogers, Matt  35  586   16.7     16.742857
1        #23 Jones, Greg  33  340   10.3     10.303030
2   #01 Stephens, Elijah  31  330   10.6     10.645161
3      #03 Smalls, Colin  33  322    9.8      9.757576
4       #33 Mayock, Matt  35  254    7.3      7.257143
5     #12 Sprouse, Geoff  20  138    6.9      6.900000
6      #04 Ball, Lincoln  35  137    3.9      3.914286
7    #07 Nausadis, Wyatt  35  107    3.1      3.057143
8     #25 Michaels, Eric  29   94    3.2      3.241379
9      #32 Iturbe, Julen  26   70    2.7      2.692308
10  #45 Whittaker, Mason  24   11    0.5      0.458333
11   #10 Samuels, Adrian  15    8    0.5      0.533333
12     #05 Daniels, Alex   3    4    1.3      1.333333
13       #20 Holod, Adam   3    3    1.0      1.000000
14       #21 Jones, Noah   7    2    0.3      0.285714

You can see our pts_per_game column is roughly equal to Pts/G — just off by rounding. 

🧑‍💻 On your own in the next cell, see if you can create a new column called "pts_per_game_rounded" that rounds your "pts_per_game" column to match AU's official one. The way to round a column to one digit is `df['column_name'].round(1)`. Use the cell above where we did division as an example.

## 🏀 Getting the data from AU basketball

*To learn Python, we are going to look at all the mens basketball games played over the last several years and see who AU has historically done well against and who they have done poorly against. This is a large, multi-step data analysis project that brings together the acquisition, cleaning and analyzing data steps that we've gone over previously in class.*

Pandas can grab data tables directly from the web. Hop over to your browser and look at the web page `https://aueagles.com/sports/mens-basketball/stats/2024-25`. If you "inspect element" you can see it's made up of tables for each stat sheet.

We want to figure out how to grab the GAME-BY-GAME table.

In [44]:
web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/2024-25")
len(web_tables)

23

The above cell gets **all** the tables from `https://aueagles.com/sports/mens-basketball/stats/2024-25`. It then prints the length `len()` of what's returned by pandas. In total, there are 23 tables on the website! 

Let's look at the first table. To access the first item in a list, you add `[0]` to the end of it. (Yes, `[0]` is 1st, don't ask why. So `[1]` is second, `[2]` is third, etc.) This is called the index of a list.

In [45]:
web_tables[0]

Statistic    American Opponents  OPP
0                            Scoring     Scoring        Scoring
1                  Total Points  TOT        2406           2370
2               Points Per Game  PPG        68.7           67.7
3             Scoring Margin  Margin         1.0             --
4                           Shooting    Shooting       Shooting
5        FG: Made-Attempted  FGM-FGA    840-1893       852-1863
6                FG: Percentage  FG%        .444           .457
7                FG: Per Game  FGM/G        24.0           24.3
8       3PT: Made-Attempted  3PT-3PA     299-859        216-652
9              3PT: Percentage  3PT%        .348           .331
10              3PT: Per Game  3PT/G         8.5            6.2
11       FT: Made-Attempted  FTM-FTA     427-566        450-643
12               FT: Percentage  FT%        .754           .700
13                FT: Per Game  FT/G        12.2           12.9
14                        Rebounding  Rebounding     Rebounding
15                        Total  TOT        1057           1129
16                     Per Game  RPG        30.2           32.3
17                            Margin        -2.1             --
18                           Assists     Assists        Assists
19                        Total  TOT         441            430
20                     Per Game  APG        12.6           12.3
21                         Turnovers   Turnovers      Turnovers
22                        Total  TOT         389            418
23                    Per Game  TO/G        11.1           11.9
24                            Margin         0.8             --
25       Assist/Turnover Ratio  A/TO         1.1            1.0
26  Points Off Turnovers  PTS OFF TO        14.4           13.0
27                            Steals      Steals         Steals
28                        Total  TOT         229            209
29                   Per Game  STL/G         6.5            6.0
30                            Blocks      Blocks         Blocks
31                        Total  TOT          66             97
32                   Per Game  BLK/G         1.9            2.8
33                        Attendance  Attendance     Attendance
34                        Total  TOT       26090          49230
35                     Per Game  AVG    16-1,631       15-3,282

🧑‍💻 See if you can find the table we want — it's the `GAME-BY-GAME` `TEAM` table. Once you've found it, store it as a variable called `scores_df`. So `scores_df = web_tables[?]`. Then output `scores_df` at the bottom of the cell.

Hint: The tables are in the same order here that they appear in the menu on AU's website, and there can be multiple tables under each tab on the website.

In [46]:
scores_df = web_tables[5]
scores_df


Opponent      Date   Score       W/L     FGM/A    PCT  \
0            at La Salle  11/04/24   52-65         L     20-55  0.364   
1             vs Harvard  11/10/24   67-55         W     22-50  0.440   
2               at Siena  11/12/24   66-74         L     24-55  0.436   
3      at North Carolina  11/15/24  55-107         L     19-61  0.311   
4          at High Point  11/18/24   73-80         L     29-52  0.558   
5         vs Kansas City  11/22/24   64-60         W     18-43  0.419   
6        at P.R.-Bayamon  11/23/24   73-56         W     26-59  0.441   
7             at UAlbany  11/24/24   81-77         W     29-45  0.644   
8     vs Notre Dame (MD)  12/01/24   97-52         W     32-59  0.542   
9   vs George Washington  12/04/24   81-71    W (OT)     25-62  0.403   
10     at Saint Joseph's  12/18/24   57-84         L     20-56  0.357   
11           at Virginia  12/22/24   58-63         L     19-50  0.380   
12               vs UMBC  12/29/24   93-96  L (2 OT)     37-77  0.481   
13               at Navy  01/02/25   58-81         L     21-56  0.375   
14         vs Holy Cross  01/05/25   75-64         W     24-48  0.500   
15          at Boston U.  01/08/25   54-60         L     18-50  0.360   
16           vs Bucknell  01/11/25   68-58         W     23-49  0.469   
17    at Loyola Maryland  01/13/25   73-54         W     28-56  0.500   
18         at Holy Cross  01/18/25   74-65         W     25-53  0.472   
19             vs Lehigh  01/22/25   68-67         W     21-46  0.457   
20            vs Colgate  01/25/25   81-77         W     23-50  0.460   
21          at Lafayette  01/29/25   75-68         W     24-59  0.407   
22    vs Army West Point  02/01/25   71-68         W     26-53  0.491   
23           at Bucknell  02/05/25   49-71         L     16-46  0.348   
24             at Lehigh  02/08/25   78-75    W (OT)     25-59  0.424   
25          vs Lafayette  02/10/25   60-58         W     23-51  0.451   
26    vs Loyola Maryland  02/15/25   72-51         W     27-52  0.519   
27    at Army West Point  02/19/25   69-76         L     26-61  0.426   
28          vs Boston U.  02/22/25   48-44         W     18-56  0.321   
29               vs Navy  02/26/25   60-68         L     21-56  0.375   
30            at Colgate  03/01/25   67-59         W     25-51  0.490   
31          vs Lafayette  03/06/25   71-69         W     25-54  0.463   
32            vs Colgate  03/09/25   72-62         W     27-54  0.500   
33               vs Navy  03/12/25   74-52         W     27-52  0.519   
34   vs Mount St. Mary's  03/19/25   72-83         L     27-57  0.474   
35              American       NaN       -       NaN  840-1893  0.444   
36             Opponents       NaN       -       NaN  852-1863  0.457   

      3FG/A  PCT.1    FTM/A  PCT.2  ...  DEF   TOT   AVG   PF  AST   TO  BLK  \
0      9-31  0.290      3-8  0.375  ...   22    31  31.0   16   13   10    3   
1      9-24  0.375    14-17  0.824  ...   27    32  31.5   21   13   11    3   
2      9-18  0.500     9-12  0.750  ...   15    24  29.0   21   11    8    2   
3      7-31  0.226    10-11  0.909  ...   16    23  27.5   23    6   13    1   
4      7-22  0.318     8-12  0.667  ...   14    26  27.2   22   13   11    1   
5      3-20  0.150    25-27  0.926  ...   23    31  27.8   19    5   20    1   
6      6-28  0.214    15-22  0.682  ...   25    33  28.6   22   14   14    3   
7      9-16  0.563    14-24  0.583  ...   20    24  28.0   25   13   16    7   
8     10-27  0.370    23-30  0.767  ...   29    42  29.6   18   19   15    2   
9     13-32  0.406    18-24  0.750  ...   30    42  30.8   18   16   16    5   
10     7-24  0.292    10-15  0.667  ...   14    25  30.3   16   11   16    2   
11    11-31  0.355      9-9  1.000  ...   23    26  29.9   14   13    7    3   
12    11-31  0.355     8-16  0.500  ...   29    41  30.8   20   21   11    4   
13     9-33  0.273     7-11  0.636  ...   21    28  30.6   18   13    9    0   
14    12-24  0.500    15-17  0.882  ...  

At this point, it's useful to think about what we need to do with the data to answer our question of total point differential per team — kind of like a recipe Let's make a set of instructions together. 

1. Read in all tables
2. Find the table that has team scores.
3. Get the score for AU and the score for the opponent.
4. Combine the points for and against each team played more than once.
5. Calculate the total point differential for each opponent.
7. Repeat for all years.

## 🔢 Pulling out the scores

Let's split out the scores so we have AU's score in one column and oponent's in another. 


In [47]:
scores_df['AU_score'] = scores_df['Score'].str.split("-",expand=True)[0]

Break that down:

`scores_df['AU_score'] = ` seting a new column alled `AU_score`

`scores_df['Score']` the score column we got from AU's website

`.str` Allows us to use some special built-in Pandas code that applies to text strings (vs numbers or date columns, for example)

`.split("-", expand=True)` "splits" the text string at the dash that separates the away and home score. Expands it so we now have a list with two numbers

`[0]` gives us the first item in that new list, which ends up being everything before the dash

How do you know how to write that? Honestly, practice. But you can always ask ChatGPT!

> **i have a score column that looks like 50-44. I want the first number pulled into a column called AU_score**

> Perfect — you can split the score string on the dash (-) and grab the first number. Here’s how in pandas:

    # Create new column with first number
    df["AU_score"] = df["score"].str.split("-").str[0]


🧑‍💻 **Now, make a new column called `opp_score` with the second number from the `Score` column**

In [49]:
scores_df['opp_score'] = scores_df['Score'].str.split("-",expand=True)[1]

In [50]:
scores_df.head(2)

Opponent      Date  Score W/L  FGM/A    PCT 3FG/A  PCT.1  FTM/A  PCT.2  \
0  at La Salle  11/04/24  52-65   L  20-55  0.364  9-31  0.290    3-8  0.375   
1   vs Harvard  11/10/24  67-55   W  22-50  0.440  9-24  0.375  14-17  0.824   

   ...   AVG  PF  AST  TO  BLK  STL  PTS  AVG.1  AU_score  opp_score  
0  ...  31.0  16   13  10    3    3   52   52.0        52         65  
1  ...  31.5  21   13  11    3    6   67   59.5        67         55  

[2 rows x 23 columns]

I'm going to use `.head()` from here on out to just look at the top of the dataframe if I want a shorter view of it. So `.head(2)` shows just the first 2 rows (plus the column headers).

If you look all the way to the right, you can see we have these two new columns. So it looks like AU lost the first game and won the second — comparing to the `W/L` column, looks correct to me! 

## ⛹️‍♂️ Standardizing the opponents

We want to tally up the points for each opponent, but right now we have that pesky "at" or "vs" in the "Opponent" column

We can get rid of it by only selecting everything after the space ... which appears to be the THIRD "character" in that column. We'll use that `.str` operator again to select everything after the third character in the column.

In [51]:
scores_df['opp_name'] = scores_df['Opponent'].str[3:]
scores_df['opp_name']

0              La Salle
1               Harvard
2                 Siena
3        North Carolina
4            High Point
5           Kansas City
6          P.R.-Bayamon
7               UAlbany
8       Notre Dame (MD)
9     George Washington
10       Saint Joseph's
11             Virginia
12                 UMBC
13                 Navy
14           Holy Cross
15            Boston U.
16             Bucknell
17      Loyola Maryland
18           Holy Cross
19               Lehigh
20              Colgate
21            Lafayette
22      Army West Point
23             Bucknell
24               Lehigh
25            Lafayette
26      Loyola Maryland
27      Army West Point
28            Boston U.
29                 Navy
30              Colgate
31            Lafayette
32              Colgate
33                 Navy
34     Mount St. Mary's
35                rican
36               onents
Name: opp_name, dtype: object

## 👋 Dropping the non-game rows

This looks good up until the last two rows. Let's look at those with the `.tail()` function — it's like the opposite of `.head()` so it shows us the bottom rows

In [83]:
scores_df.tail()

Opponent      Date  Score  W/L     FGM/A    PCT    3FG/A  \
32           vs Colgate  03/09/25  72-62    W     27-54  0.500     7-17   
33              vs Navy  03/12/25  74-52    W     27-52  0.519    11-25   
34  vs Mount St. Mary's  03/19/25  72-83    L     27-57  0.474    13-35   
35             American       NaN      -  NaN  840-1893  0.444  299-859   
36            Opponents       NaN      -  NaN  852-1863  0.457  216-652   

    PCT.1    FTM/A  PCT.2  ...   PF  AST   TO  BLK  STL   PTS  AVG.1  \
32  0.412    11-13  0.846  ...    9   11   11    1    8    72   68.5   
33  0.440     9-13  0.692  ...   13   15    9    3    8    74   68.6   
34  0.371      5-8  0.625  ...   18   15   13    3    4    72   68.7   
35  0.348  427-566  0.754  ...  625  441  389   66  229  2406   68.7   
36  0.331  450-643  0.700  ...  548  430  418   97  209  2370   67.7   

    AU_score  opp_score          opp_name  
32        72         62           Colgate  
33        74         52              Navy  
34        72         83  Mount St. Mary's  
35                                  rican  
36                                 onents  

[5 rows x 24 columns]

For these two rows, they appear to be totals so we should "drop" them from the dataframe. 

HOWEVER, we want to do this in a way that's ideally automatable. So we don't just want to take row number 35 and 36, since maybe in a previous year those totals don't exist, or they are on different row numbers. 

Instead, let's "filter" the dataframe to just rows where the date is not blank. 

See how the date on those rows is `NaN`? That's pandas-speak for "not a number". We can filter those out like this:

In [54]:
scores_df = scores_df[scores_df['Date'].isna()==False]

In general, the way you filter a pandas dataframe to specific rows is like this:

`df[df['column_name'] == X]` => filters to rows where a column is equal to X

`df[df['column_name'] != X]` => filters to rows where a column is NOT equal to X

`df[df['column_name'] > X]` => filters to rows where a column is greater than X

`df[df['column_name'].isna()]` => filters to rows where a column is blank

In our case, we're looking for where a column is NOT blank. So we say `.isna()==False`, which checks if the column is null, and returns False if so.

In [55]:
scores_df.tail()

Opponent      Date  Score W/L  FGM/A    PCT  3FG/A  PCT.1  \
30           at Colgate  03/01/25  67-59   W  25-51  0.490  12-26  0.462   
31         vs Lafayette  03/06/25  71-69   W  25-54  0.463  10-23  0.435   
32           vs Colgate  03/09/25  72-62   W  27-54  0.500   7-17  0.412   
33              vs Navy  03/12/25  74-52   W  27-52  0.519  11-25  0.440   
34  vs Mount St. Mary's  03/19/25  72-83   L  27-57  0.474  13-35  0.371   

    FTM/A  PCT.2  ...  PF  AST  TO  BLK  STL  PTS  AVG.1  AU_score  opp_score  \
30    5-8  0.625  ...  12   16  13    1    8   67   68.3        67         59   
31  11-16  0.688  ...  10   13  11    1   10   71   68.4        71         69   
32  11-13  0.846  ...   9   11  11    1    8   72   68.5        72         62   
33   9-13  0.692  ...  13   15   9    3    8   74   68.6        74         52   
34    5-8  0.625  ...  18   15  13    3    4   72   68.7        72         83   

            opp_name  
30           Colgate  
31         Lafayette  
32           Colgate  
33              Navy  
34  Mount St. Mary's  

[5 rows x 24 columns]

Now the total rows are gone when we look at the tail! 

## ➕ Combining for each opponent

So now we can add up the total AU points and total opponent points for each opponent. For an opponent like "Navy" we should see totals added together, since we played them 3 times.

In [89]:
scores_combined_df = scores_df.groupby("opp_name").sum()[['AU_score','opp_score']]

`scores_combined_df = ` creates a brand new dataframe (think of it like a new pivot table or new tab in a spreadsheet)

`scores_df.groupby("opp_name")` groups the scores_df dataframe into one row per value in `opp_name` column (one row per opponent) 

`.sum()` adds up the grouped values in every column

`[['AU_score','opp_score']]` only returns the AU_score and opponent score columns

In [91]:
scores_combined_df

AU_score opp_score
opp_name                            
Army West Point       7169      6876
Boston U.             5448      6044
Bucknell              6849      5871
Colgate             816772    775962
George Washington       81        71
Harvard                 67        55
High Point              73        80
Holy Cross            7574      6465
Kansas City             64        60
La Salle                52        65
Lafayette           756071    685869
Lehigh                6878      6775
Loyola Maryland       7372      5451
Mount St. Mary's        72        83
Navy                586074    816852
North Carolina          55       107
Notre Dame (MD)         97        52
P.R.-Bayamon            73        56
Saint Joseph's          57        84
Siena                   66        74
UAlbany                 81        77
UMBC                    93        96
Virginia                58        63

But wait! Something loks wrong. Did AU really score 586,074 points against Navy? 

Nope. 

This is "summing" *strings*, not integers, together. So 58 AU's first game, 60 from AU's second game and 74 from AU's third game ...

We need to tell Pandas that that column are actually *integers*. Not strings of text.

So we here we go back to our original scores_df dataframe and recast certain columns as integers (ignore the warnings that pop up.)

In [56]:
scores_df['AU_score'] = scores_df['AU_score'].astype(int)

🧑‍💻 **Do the same for opp_score**

In [57]:
scores_df['opp_score'] = scores_df['opp_score'].astype(int)

Now we'll go back and create our groupby dataframe again, and this time it should know those 2 columns are integers that need to be summed

In [58]:
scores_combined_df = scores_df.groupby("opp_name").sum()[['AU_score','opp_score']]
scores_combined_df


AU_score  opp_score
opp_name                              
Army West Point         140        144
Boston U.               102        104
Bucknell                117        129
Colgate                 220        198
George Washington        81         71
Harvard                  67         55
High Point               73         80
Holy Cross              149        129
Kansas City              64         60
La Salle                 52         65
Lafayette               206        195
Lehigh                  146        142
Loyola Maryland         145        105
Mount St. Mary's         72         83
Navy                    192        201
North Carolina           55        107
Notre Dame (MD)          97         52
P.R.-Bayamon             73         56
Saint Joseph's           57         84
Siena                    66         74
UAlbany                  81         77
UMBC                     93         96
Virginia                 58         63

Better! 

So now we have the total for and against each opponent, but we really want the point differential. 

Just like we used division when we calculated points per game earlier on, we'll do the same here by creating a new column that subtracts opponent score from AU score. 

🧑‍💻 **Create a new column in `scores_combined_df` called `point_diff` that is equal to `AU_score` - `opp_score`**

In [59]:
scores_combined_df['points_diff'] = scores_combined_df['AU_score'] - scores_combined_df['opp_score']

In [60]:
scores_combined_df

AU_score  opp_score  points_diff
opp_name                                           
Army West Point         140        144           -4
Boston U.               102        104           -2
Bucknell                117        129          -12
Colgate                 220        198           22
George Washington        81         71           10
Harvard                  67         55           12
High Point               73         80           -7
Holy Cross              149        129           20
Kansas City              64         60            4
La Salle                 52         65          -13
Lafayette               206        195           11
Lehigh                  146        142            4
Loyola Maryland         145        105           40
Mount St. Mary's         72         83          -11
Navy                    192        201           -9
North Carolina           55        107          -52
Notre Dame (MD)          97         52           45
P.R.-Bayamon             73         56           17
Saint Joseph's           57         84          -27
Siena                    66         74           -8
UAlbany                  81         77            4
UMBC                     93         96           -3
Virginia                 58         63           -5

Last thing, let's sort the dataframe by that new column to see who AU did best and worst against overall

In [61]:
scores_combined_df.sort_values("points_diff")

AU_score  opp_score  points_diff
opp_name                                           
North Carolina           55        107          -52
Saint Joseph's           57         84          -27
La Salle                 52         65          -13
Bucknell                117        129          -12
Mount St. Mary's         72         83          -11
Navy                    192        201           -9
Siena                    66         74           -8
High Point               73         80           -7
Virginia                 58         63           -5
Army West Point         140        144           -4
UMBC                     93         96           -3
Boston U.               102        104           -2
Kansas City              64         60            4
Lehigh                  146        142            4
UAlbany                  81         77            4
George Washington        81         71           10
Lafayette               206        195           11
Harvard                  67         55           12
P.R.-Bayamon             73         56           17
Holy Cross              149        129           20
Colgate                 220        198           22
Loyola Maryland         145        105           40
Notre Dame (MD)          97         52           45

## 🧑‍🤝‍🧑 Combining all that 

What we've done so far is 

* Get the 2024 table from the AU website
* Pull out the scores
* Standardize the opponents
* Drop the non-game total rows
* Combine the scores for each opponent

We've done this across several cells, but all in one cell it could look like this (again, ignore the warnings):

In [62]:
import pandas as pd

# get the data from AU
web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/2024-25")
scores_df = web_tables[5]

# pull out the scores
scores_df['AU_score'] = scores_df['Score'].str.split("-",expand=True)[0]
scores_df['opp_score'] = scores_df['Score'].str.split("-",expand=True)[1]

# standardize opponent name by removing vs or at
scores_df['opp_name'] = scores_df['Opponent'].str[3:]

# drop the non game total rows
scores_df = scores_df[scores_df['Date'].isna()==False]

# combine for each opponent
scores_df['AU_score'] = scores_df['AU_score'].astype(int)
scores_df['opp_score'] = scores_df['opp_score'].astype(int)
scores_combined_df = scores_df.groupby("opp_name").sum()[['AU_score','opp_score']]
scores_combined_df['points_diff'] = scores_combined_df['AU_score'] - scores_combined_df['opp_score']

# output the sorted combined scores dataframe
scores_combined_df.sort_values("points_diff")


/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_11565/3656780976.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_df['AU_score'] = scores_df['AU_score'].astype(int)
/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_11565/3656780976.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_df['opp_score'] = scores_df['opp_score'].astype(int)


AU_score  opp_score  points_diff
opp_name                                           
North Carolina           55        107          -52
Saint Joseph's           57         84          -27
La Salle                 52         65          -13
Bucknell                117        129          -12
Mount St. Mary's         72         83          -11
Navy                    192        201           -9
Siena                    66         74           -8
High Point               73         80           -7
Virginia                 58         63           -5
Army West Point         140        144           -4
UMBC                     93         96           -3
Boston U.               102        104           -2
Kansas City              64         60            4
Lehigh                  146        142            4
UAlbany                  81         77            4
George Washington        81         71           10
Lafayette               206        195           11
Harvard                  67         55           12
P.R.-Bayamon             73         56           17
Holy Cross              149        129           20
Colgate                 220        198           22
Loyola Maryland         145        105           40
Notre Dame (MD)          97         52           45

Note the comments there, which aren't telling Python to do anything but just makes it more readable for us humans

## 🗓 Putting in other years of data

So somewhere in here we want to make this automated, where it goes and pulls down all the data. **Discussion: How could this change the code above?**

### 🏀 Scraping the other years from AU's website

Notice the url that we pulled the data from `https://aueagles.com/sports/mens-basketball/stats/2024-25`

Those two years on the end of it look interesting. What would happen if we changed that? **Test that in your browser to try to go back to an earlier year (ie 2021-2022)**

Really what we want to do is go through *every* year available and grab that score table on the page. 

Then we want to put it into one GIANT score table with every game from every year. 

Looking at the dropdown on the AU web page, it looks like the first season they have is 2009-10. 

So we want to write a list of years that we'll get the data from:

`years_list = ['2009-10', '2010-11', ..., '2024-25']`

🧑‍💻 **Finish typing that years_list variable in the cell below, making sure each item in the list is a string surrounded by quotes**

In [63]:
years_list = ["2009-10", "2010-11", "2011-12", "2012-13", "2013-14", "2014-15", "2015-16", "2016-17", "2017-18", "2018-19", "2019-20", "2020-21", "2021-22", "2022-23", "2023-24", '2024-25']


Now we can go through each of those years and get the score table for that year. Along the way, we'll add it to one GIANT score dataframe that will start empty but get filled in for each year.

In [64]:
giant_score_df = pd.DataFrame()

In [65]:
import time

for year_string in years_list:
    web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/" + year_string)
    scores_df = web_tables[5]
    giant_score_df = pd.concat([giant_score_df,scores_df])
    print(year_string, 'completed')
    time.sleep(4)
    
    

2009-10 completed
2010-11 completed
2011-12 completed
2012-13 completed
2013-14 completed


HTTPError: HTTP Error 404: Not Found

Uh oh! That error looks scary. 

But before we dive into it, let's go line by line and look at what that code was doing:

    import time

    for year_string in years_list:
        web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/" + year_string)
        scores_df = web_tables[5]
        giant_score_df = pd.concat([giant_score_df,scores_df])
        print(year_string, 'completed')
        time.sleep(4)
        


        


`import time` imports a new library called time that lets us pause for a few seconds between each request to AU's website 

`for year_string in years_list:` sets up a "for loop". this is like saying "for x in y" where y is a list and x is what we're going to call each item in the list as we go through it. The "x" name is arbitrary — we could say `for banana in years_list` and it would still work
   
`web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/" + year_string)` the same as when we got the web_tables before, except this time we are using `year_string` — the "x in y". So in years_list, `year_string` will represent either "2009-10", or "2010-11", etc...

`scores_df = web_tables[5]` same as before, gets the sixth web table (since index 5 = the 6th). This only holds one year at a time.

`giant_score_df = pd.concat([giant_score_df,scores_df])` uses pandas to combine, or *concatenate*, two dataframes: `giant_score_df` and `scores_df`. This is how we keep adding on and on to `giant_score_df`

`print(year_string, 'completed')` prints the year then the word "completed", letting us know when we've finished a year in our for loop

`time.sleep(4)` pauses for four seconds so we dont overwhelm the servers from the website we're scraping

-------

So now that we know how that for loop works, we can look at the error. Always start from the bottom of the error - in this one, we see there's a 404 error for the one of the pages.

The error appeared after we finished the 2013-2014 season, which means it's probably a problem with the 2014-15 sesason. **Go to the AU web page in your browser and see if you can figure out what's going on.**

------------------

It looks like 2014-15 goes to a broken link, so we'll have to skip that year. Here's the `years_list` rewritten without 2014-15 included

In [67]:
years_list = ["2009-10", "2010-11", "2011-12", "2012-13", "2013-14", "2015-16", "2016-17", "2017-18", "2018-19", "2019-20", "2020-21", "2021-22", "2022-23", "2023-24", '2024-25']


After you run this cell above, try re-running the for loop here

In [68]:
giant_score_df = pd.DataFrame()

for year_string in years_list:
    web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/" + year_string)
    scores_df = web_tables[5]
    giant_score_df = pd.concat([giant_score_df,scores_df])
    print(year_string, 'completed')
    time.sleep(4)

2009-10 completed
2010-11 completed
2011-12 completed
2012-13 completed
2013-14 completed
2015-16 completed
2016-17 completed
2017-18 completed
2018-19 completed
2019-20 completed
2020-21 completed
2021-22 completed
2022-23 completed
2023-24 completed
2024-25 completed


OK, looks like all the rest worked! We can now view the giant dataframe and see what it looks like 

In [69]:
giant_score_df

Opponent      Date  Score  W/L     FGM/A    PCT    3FG/A  \
0   at SAINT FRANCIS-PA  11/13/09  61-65    L     22-52  0.423     3-12   
1             vs ALBANY  11/16/09  50-51    L     21-55  0.382     3-14   
2              vs MOUNT  11/18/09  62-73    L     24-59  0.407     6-14   
3                vs FAU  11/21/09  69-81    L     23-57  0.404     3-14   
4                vs FFD  11/23/09  56-72    L     22-50  0.440     7-15   
..                  ...       ...    ...  ...       ...    ...      ...   
32           vs Colgate  03/09/25  72-62    W     27-54  0.500     7-17   
33              vs Navy  03/12/25  74-52    W     27-52  0.519    11-25   
34  vs Mount St. Mary's  03/19/25  72-83    L     27-57  0.474    13-35   
35             American       NaN      -  NaN  840-1893  0.444  299-859   
36            Opponents       NaN      -  NaN  852-1863  0.457  216-652   

    PCT.1    FTM/A  PCT.2  ...  DEF   TOT   AVG   PF  AST   TO  BLK  STL  \
0   0.250    14-15  0.933  ...   23    32  32.0   20   12   13    0    5   
1   0.214      5-5  1.000  ...   24    32  32.0   10   13   11    2    4   
2   0.429     8-12  0.667  ...   26    40  34.7   21   12   12    4    4   
3   0.214    20-35  0.571  ...   28    37  35.3   24   18   13    3    6   
4   0.467     5-13  0.385  ...   23    37  35.6   15   15   20    3    4   
..    ...      ...    ...  ...  ...   ...   ...  ...  ...  ...  ...  ...   
32  0.412    11-13  0.846  ...   25    32  30.3    9   11   11    1    8   
33  0.440     9-13  0.692  ...   27    31  30.3   13   15    9    3    8   
34  0.371      5-8  0.625  ...   18    26  30.2   18   15   13    3    4   
35  0.348  427-566  0.754  ...  787  1057  30.2  625  441  389   66  229   
36  0.331  450-643  0.700  ...  834  1129  32.3  548  430  418   97  209   

     PTS  AVG.1  
0     61   61.0  
1     50   55.5  
2     62   57.7  
3     69   60.5  
4     56   59.6  
..   ...    ...  
32    72   68.5  
33    74   68.6  
34    72   68.7  
35  2406   68.7  
36  2370   67.7  

[479 rows x 21 columns]

Not bad, 479 rows!

Here's a couple things we need to make sure of before we put this into our working code:

1. Need to get rid of totals
2. Need to make sure that schools are written the same over time (this could be hard since we're going back 10+ years)

🧑‍💻 Let's start with the easy one, get rid of the totals, aka the rows that have a blank `Date` column. **Filter this down in the next cell as a NEW variable called `giant_score_df_filtered` with only the rows of `giant_score_df` where Date is not blank** (so we don't overwrite the dataframe we just worked so hard to scrape)

In [71]:
giant_score_df_filtered = giant_score_df[giant_score_df['Date'].notna()]

🧑‍💻 **Once you've done that, print out the length of the new dataframe: `print(len(giant_score_df_filtered))`. See how it compares to the original before you filtered it `print(len(giant_score_df))`**

In [72]:
print(len(giant_score_df_filtered))
len(giant_score_df)


449


479

It should be that there are 30 rows that were dropped - this makes sense if we have 15 years worth of data, each with two total rows! (one for AU and one for opponents)

### ⛹️‍♂️ Standardize school names 

Now we want to look at how school names are written, since we can't do a `groupby` if they're different.

🧑‍💻 First, put the names into `opp_name` column again so we can get rid of the "at" and "vs" in the opponent column. This will be the same as how we did it previously.

In [73]:
giant_score_df_filtered['opp_name'] = giant_score_df_filtered['Opponent'].str[3:]

/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_11565/1969781723.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  giant_score_df_filtered['opp_name'] = giant_score_df_filtered['Opponent'].str[3:]


Now I want to look at all the UNQIUE names in `opp_name` column and see if there are schools that are written more than one way

Outputting a pandas list into a `set()` shows us the unique entries in it. 

In [74]:
set(giant_score_df_filtered['opp_name'].sort_values())

{'AKRON',
 'ALBANY',
 'ARMY',
 'Albany (NY)',
 'Army',
 'Army West Point',
 'BRWN',
 'BUCKNELL',
 'Boston U',
 'Boston U.',
 'Boston University',
 'Brown',
 'Bucknell',
 'Buffalo',
 'COL',
 'COLGATE',
 'COLGATEM',
 'CORNELL',
 'CU',
 'Colgate',
 'Columbia',
 'Cornell',
 'DePaul',
 'Delaware',
 'Duquesne',
 'FAU',
 'FFD',
 'Florida',
 'Florida Atlantic',
 'Fordham',
 'GEORGETOWN',
 'George Mason',
 'George Washington',
 'Georgetown',
 'HC',
 'HOLY CROSS',
 'HOW-M',
 'HOWM',
 'Hampton',
 'Harvard',
 'High Point',
 'Holy Cross',
 'Hood',
 'Howard',
 'Howard University',
 'Kansas',
 'Kansas City',
 'Kansas State',
 'LAFAYETT',
 'LAFAYETTE',
 'LEHIGH',
 'LSU',
 'La Salle',
 'Lafayette',
 'Lehigh',
 'Longwood',
 'Loyola',
 'Loyola Maryland',
 'MANHATTAN',
 'MOUNT',
 'Marist',
 'Marquette',
 'Maryland Eastern Shore',
 'Maryland Terrapins',
 'Md. Eastern Shore',
 'Minnesota',
 "Mount St. Mary's",
 'NAVY',
 'NJIT',
 'Navy',
 'New Hampshire',
 'North Carolina',
 'Northwestern',
 'Notre Dame (MD)

Right off the bat we can see issues. Albany is in there as "ALBANY" and "Albany (NY)". Boston University looks like it's in there 3 different ways.

----------

To account for this, we can set up a *dictionary* that tells Pandas how we want each to be referred to. A dictionary is like a list, except instead of referring to values in it by an index ([0], [1], etc...), we refer to them by a *key*. 

So `sean_dict = {"hair": "brown", "name": "sean", "eyes": "blue"}` could be a dictionary representing me. 

If we output `sean_dict['hair']`, we would expect it to return "brown"

---------------------

🧑‍💻 Here, we'll start a dictionary that says how we want to refer to the ones that need to change. **We'll go through and fill it in together.**

In [77]:
school_names_dict = {
    'Albany (NY)': "ALBANY",
    'ARMY': "Army",
    'Army West Point': "Army"
}


Once we have that done, we'll want to make that a new column called `opp_name_cleaned`. We can change the names in `opp_name` column that are keys in our dictionary. 

In [78]:
def replace_name(x):
    if x in school_names_dict:
        return school_names_dict[x]
    else:
        return x

giant_score_df_filtered['opp_name_cleaned'] = giant_score_df_filtered['opp_name'].apply(replace_name)

/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_11565/305087932.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  giant_score_df_filtered['opp_name_cleaned'] = giant_score_df_filtered['opp_name'].apply(replace_name)


`def replace_name(x)` is a *function*. It takes an input (x) and then returns an output depending on what x is.

`if x in school_names_dict:` says if the input `x` exists as a key in our dictionary, do the following:

`return school_names_dict[x]` would in that case return the *value* of that dictionary key `x`

`else:` says if `x` does NOT exist as a key in our dictionary, do the following:

`return x` just returns the input `x` without doing anything to it. 

In total, this function either substitutes in a "clean" name or just returns the original name. But this above function doesn't "automatically" run. We have to tell Python or Pandas how to use it below:

`giant_score_df_filtered['opp_name_cleaned'] =` sets new column called `opp_name_cleaned`

`giant_score_df_filtered['opp_name']` accesses our old column called `opp_name`

`.apply(replace_name)` applies our `replace_name` function to that column 



-------------

Let's look at the rows that changed - the ones where `opp_name_cleaned` is not the same as `opp_name`

In [79]:
giant_score_df_filtered[giant_score_df_filtered['opp_name_cleaned'] !=  giant_score_df_filtered['opp_name']]

Opponent      Date  Score W/L  FGM/A    PCT  3FG/A  PCT.1  \
25             vs ARMY  02/17/10  57-54   W  20-44  0.455  10-21  0.476   
18             vs ARMY  01/19/11  65-62   W  22-46  0.478   6-15  0.400   
19             vs ARMY  01/21/12  67-55   W  25-53  0.472   9-18  0.500   
29             vs ARMY  02/29/12  57-40   W  18-46  0.391   6-17  0.353   
26             vs ARMY  02/22/14  58-54   W  21-42  0.500   5-12  0.417   
16  at Army West Point  01/17/16  45-65   L  17-47  0.362   3-16  0.188   
24             vs ARMY  02/13/16  59-65   L  21-57  0.368   9-24  0.375   
16             vs ARMY  01/14/17  49-53   L  22-57  0.386   4-20  0.200   
24             at ARMY  02/11/17  61-58   W  21-46  0.457   5-10  0.500   
29             at ARMY  02/28/17  58-74   L  21-57  0.368   7-23  0.304   
14  vs Army West Point  01/08/18  54-58   L  20-55  0.364   6-16  0.375   
22  at Army West Point  02/03/18  70-82   L  26-54  0.481   7-22  0.318   
22             vs ARMY  02/09/19  71-68   W  27-54  0.500   4-16  0.250   
26             at ARMY  02/24/19  77-66   W  30-58  0.517   7-23  0.304   
5       at Albany (NY)  11/30/19  64-68   L  25-67  0.373   8-26  0.308   
13  at Army West Point  01/08/20  68-60   W  24-54  0.444   7-21  0.333   
22  vs Army West Point  02/08/20  72-62   W  24-61  0.393  10-25  0.400   
9   at Army West Point  03/06/21  66-89   L  25-53  0.472  11-26  0.423   
15  at Army West Point  01/19/22  66-76   L  27-54  0.500   8-20  0.400   
26  vs Army West Point  02/19/22  83-67   W  35-60  0.583   6-16  0.375   
15  at Army West Point  01/11/23  60-72   L  22-49  0.449   4-19  0.211   
19  vs Army West Point  01/25/23  63-61   W  25-51  0.490   4-14  0.286   
16  vs Army West Point  01/13/24  79-60   W  32-55  0.582   8-18  0.444   
29  at Army West Point  02/28/24  73-51   W  27-48  0.563  12-25  0.480   
22  vs Army West Point  02/01/25  71-68   W  26-53  0.491   9-25  0.360   
27  at Army West Point  02/19/25  69-76   L  26-61  0.426   9-26  0.346   

    FTM/A  PCT.2  ...   AVG  PF  AST  TO  BLK  STL  PTS  AVG.1  \
25   7-11  0.636  ...  35.3  16   16  17    1    1   57   62.0   
18  15-29  0.517  ...  33.3  18   14  10    5    5   65   65.3   
19   8-10  0.800  ...  32.1  15   16  13    3    5   67   64.7   
29  15-20  0.750  ...  32.6  17   14  14    3    5   57   64.4   
26  11-15  0.733  ...  30.1  25   15  16    2   11   58   65.2   
16   8-12  0.667  ...  27.6  20    9  12    1    5   45   52.9   
24   8-11  0.727  ...  27.3  18   13   9    1    6   59   55.4   
16    1-4  0.250  ...  30.5  11   14   9    1    6   49   61.5   
24  14-21  0.667  ...  28.8  21   16  15    4   10   61   61.2   
29   9-12  0.750  ...  28.9  15   11   9    2   10   58   61.5   
14   8-16  0.500  ...  29.7  15   12  16    4   15   54   63.6   
22  11-18  0.611  ...  29.0  14   11  11    5    3   70   63.4   
22  13-17  0.765  ...  34.2  20   16  13    9    6   71   71.4   
26  10-15  0.667  ...  33.9  11   12  14    6    8   77   72.3   
5    6-10  0.600  ...  31.8  18   16  13    2    8   64   73.8   
13  13-16  0.813  ...  32.4   7   15  10    2    8   68   70.2   
22  14-18  0.778  ...  32.9  15   11  11    7    7   72   71.3   
9     5-9  0.556  ...  32.6  16   12  15    1    6   66   69.1   
15    4-8  0.500  ...  30.9  21   15  17    1    9   66   64.4   
26   7-10  0.700  ...  30.3  15   16  13    4    9   83   63.8   
15  12-19  0.632  ...  30.6  19    8   9    2   10   60   68.1   
19   9-10  0.900  ...  30.2  13   11  14    7    7   63   66.8   
16   7-12  0.583  ...  34.2  21   21  12    0    5   79   73.5   
29    7-8  0.875  ...  32.8  12   23  10    1    6   73   69.2   
22  10-16  0.625  ...  30.6  16   12  15    1    4   71   70.2   
27   8-13  0.615  ...  30.4  19   17   7    1    7   69   69.4   

           opp_name  opp_name_cleaned  
25             ARMY              Army  
18             ARMY              Army  
19             ARMY              Army  
29             ARMY              Army  
26   

Now that we have our giant_score_df_filtered, we can put this code into our old Python code block to integrate it. 



### ✅ A minimum threshold of times playing a team 

But wait, there's one more thing!

AU played some teams a bunch of times and some teams very few times over this 15 year period. Maybe we want to set a minimum threshold? So we only look at teams that we played at least X times. 

First, we get the total number of times we played each team and save it as a variable `counts`

In [80]:
counts = giant_score_df_filtered['opp_name_cleaned'].value_counts()
counts

opp_name_cleaned
Army               32
Navy               31
Bucknell           28
Lehigh             27
Holy Cross         27
                   ..
High Point          1
Kansas City         1
P.R.-Bayamon        1
Notre Dame (MD)     1
Virginia            1
Name: count, Length: 110, dtype: int64

`.value_counts()` is a build in pandas function that tells us how many times each value shows up in a column

Now, set the minimum threshold we want to work with as a variable, so it's easy to change later if we want to play with it 

In [84]:
min_threshold = 3


Then, keep only rows where the opponent appears at least min_threshold times and store as a new dataframe variable `giant_score_df_filtered_cts`

In [85]:
giant_score_df_filtered_cts = giant_score_df_filtered[
    giant_score_df_filtered["opp_name_cleaned"].map(counts) >= min_threshold
]

`.map(counts)` says: “Look up the count for each row’s opponent.” in our variable `counts`

`>= min_threshold` only keep the rows that are at least as big as the number we set as min_threshold (greater than or equal to)

🧑‍💻 **Again using `len()`, see how many rows we just dropped between `giant_score_df_filtered_cts` and `giant_score_df_filtered`**

In [86]:
len(giant_score_df_filtered_cts)

354

In [87]:
len(giant_score_df_filtered)

449

## 🧑‍🤝‍🧑🗓 Putting it all together, with all the years of data 

As a reminder, here's the old code block we wrote

In [88]:
import pandas as pd

# get the data from AU
web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/2024-25")
scores_df = web_tables[5]

# pull out the scores
scores_df['AU_score'] = scores_df['Score'].str.split("-",expand=True)[0]
scores_df['opp_score'] = scores_df['Score'].str.split("-",expand=True)[1]

# standardize opponent name by removing vs or at
scores_df['opp_name'] = scores_df['Opponent'].str[3:]

# drop the non game total rows
scores_df = scores_df[scores_df['Date'].isna()==False]

# combine for each opponent
scores_df['AU_score'] = scores_df['AU_score'].astype(int)
scores_df['opp_score'] = scores_df['opp_score'].astype(int)
scores_combined_df = scores_df.groupby("opp_name").sum()[['AU_score','opp_score']]
scores_combined_df['points_diff'] = scores_combined_df['AU_score'] - scores_combined_df['opp_score']

# output the sorted combined scores dataframe
scores_combined_df.sort_values("points_diff")


/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_11565/3656780976.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_df['AU_score'] = scores_df['AU_score'].astype(int)
/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_11565/3656780976.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_df['opp_score'] = scores_df['opp_score'].astype(int)


AU_score  opp_score  points_diff
opp_name                                           
North Carolina           55        107          -52
Saint Joseph's           57         84          -27
La Salle                 52         65          -13
Bucknell                117        129          -12
Mount St. Mary's         72         83          -11
Navy                    192        201           -9
Siena                    66         74           -8
High Point               73         80           -7
Virginia                 58         63           -5
Army West Point         140        144           -4
UMBC                     93         96           -3
Boston U.               102        104           -2
Kansas City              64         60            4
Lehigh                  146        142            4
UAlbany                  81         77            4
George Washington        81         71           10
Lafayette               206        195           11
Harvard                  67         55           12
P.R.-Bayamon             73         56           17
Holy Cross              149        129           20
Colgate                 220        198           22
Loyola Maryland         145        105           40
Notre Dame (MD)          97         52           45

And here it is, after we  .... 

* scraped all the years from AU
* cleaned school names
* set a minimum threshold
* changed variable names to match our giant_score_df instead of just one year at a time

(Note this will take a little bit to run as it will "re-scrape" each of those URLs)

In [19]:
import pandas as pd

# *NEW* Scrape all the years from AU
years_list = ["2009-10", "2010-11", "2011-12", "2012-13", "2013-14", "2015-16", "2016-17", "2017-18", "2018-19", "2019-20", "2020-21", "2021-22", "2022-23", "2023-24", '2024-25']
giant_score_df = pd.DataFrame()

for year_string in years_list:
    web_tables = pd.read_html("https://aueagles.com/sports/mens-basketball/stats/" + year_string)
    scores_df = web_tables[5]
    giant_score_df = pd.concat([giant_score_df,scores_df])
    print(year_string, 'completed')
    time.sleep(4)

# standardize opponent name by removing vs or at (*moved up before we fix school names*)
giant_score_df['opp_name'] = giant_score_df['Opponent'].str[3:]

# drop the non game total rows (*moved up before we fix school names*)
giant_score_df_filtered = giant_score_df[giant_score_df['Date'].isna()==False]

# *NEW* fix school names that are written differently
school_names_dict = {
    'Albany (NY)': "ALBANY",
    'ARMY': "Army",
    'Army West Point': "Army"
}
giant_score_df_filtered['opp_name_cleaned'] = giant_score_df_filtered['opp_name'].apply(lambda x: school_names_dict[x] if x in school_names_dict else x)

# *NEW* set a minimum threshold of how many times AU has to play a team to be counted
counts = giant_score_df_filtered['opp_name_cleaned'].value_counts()
min_threshold = 3
giant_score_df_filtered_cts = giant_score_df_filtered[
    giant_score_df_filtered["opp_name_cleaned"].map(counts) >= min_threshold
]

# *NEW* from here on out we are referring to giant_score_df_filtered_cts to get all the scores, which replaces scores_df

# pull out the scores
giant_score_df_filtered_cts['AU_score'] = giant_score_df_filtered_cts['Score'].str.split("-",expand=True)[0]
giant_score_df_filtered_cts['opp_score'] = giant_score_df_filtered_cts['Score'].str.split("-",expand=True)[1]

# combine for each opponent
giant_score_df_filtered_cts['AU_score'] = giant_score_df_filtered_cts['AU_score'].astype(int)
giant_score_df_filtered_cts['opp_score'] = giant_score_df_filtered_cts['opp_score'].astype(int)
scores_combined_df = giant_score_df_filtered_cts.groupby("opp_name_cleaned").sum()[['AU_score','opp_score']] # *NEW* refers to opp_name_cleaned instead of opp_name
scores_combined_df['points_diff'] = scores_combined_df['AU_score'] - scores_combined_df['opp_score']

# output the sorted combined scores dataframe
scores_combined_df.sort_values("points_diff")


2009-10 completed
2010-11 completed
2011-12 completed
2012-13 completed
2013-14 completed
2015-16 completed
2016-17 completed
2017-18 completed
2018-19 completed
2019-20 completed
2020-21 completed
2021-22 completed
2022-23 completed
2023-24 completed
2024-25 completed


/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_13223/4029434657.py:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  giant_score_df_filtered['opp_name_cleaned'] = giant_score_df_filtered['opp_name'].apply(lambda x: school_names_dict[x] if x in school_names_dict else x)
/var/folders/4p/sk1kyy7d621cjpk6t3v7dxp00000gn/T/ipykernel_13223/4029434657.py:38: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  giant_score_df_filtered_cts['AU_score'] = giant_score_df_filtered_cts['Score'].str.split("-",expand=T

AU_score  opp_score  points_diff
opp_name_cleaned                                    
Lehigh                  1766       1918         -152
Bucknell                1774       1897         -123
Georgetown               375        471          -96
Villanova                163        253          -90
Boston U.                649        721          -72
Colgate                 1594       1666          -72
Saint Joseph's           165        227          -62
George Mason             291        346          -55
Navy                    1929       1978          -49
Stony Brook              199        243          -44
BUCKNELL                 275        315          -40
Saint Francis            171        203          -32
LEHIGH                   254        280          -26
Lafayette               1828       1853          -25
Columbia                 206        229          -23
Mount St. Mary's         705        728          -23
Hampton                  187        205          -18
William & Mary           271        280           -9
Siena                    285        292           -7
Army                    2055       2061           -6
Boston University        652        647            5
ALBANY                   171        163            8
SFU                      212        203            9
Howard                   492        482           10
Brown                    213        201           12
George Washington        284        271           13
Saint Francis (PA)       298        282           16
UMBC                     733        710           23
UMES                     209        186           23
LAFAYETTE                281        251           30
VMI                      289        258           31
CU                       211        176           35
New Hampshire            280        244           36
NAVY                     406        336           70
Loyola Maryland         1474       1404           70
Holy Cross              1782       1645          137

This table above reflects the total points for and against teams that AU played at least three times between the 2010 and 2025 seasons (excluding the 2015 season, which had no data).

# 📈 Using Datawrapper to automatically visualize this 

## 📚 Loading in the library

Now let's look at how we would automatically put this into a chart using Datawrapper's API. First thing we need to do is install the `datawrapper` library and import it 

In [90]:
%pip install datawrapper

Note: you may need to restart the kernel to use updated packages.


In [2]:
import datawrapper as dw

Datawrapper's API, like most do, requires a secret key so that it knows who is using it. Think of it like a password. There are [instructions here on how to do that](https://developer.datawrapper.de/docs/getting-started) under Authentication.

We're going to set up a Python Datawrapper object that "knows" our API key. 

## 🔑 Setting up the API 

🧑‍💻 Insert your API key below to the right of the equal sign

In [1]:
%env DATAWRAPPER_ACCESS_TOKEN=

env: DATAWRAPPER_ACCESS_TOKEN=


Again here, nothing "happens" because all we've done is name the `dw` variable as this object. 

BUT — now we have access to all the different fun things we can do with a Datawrapper object, known as "methods." Methods are the prebuilt functions that come with libraries that chain off an object like `pandas.sort_values()`. 



## 🐣 Creating a chart

Most good libraries and APIs have what's called "documentation" online where they write how to use the library or API.

Here is the documentation for Datawrapper's API https://datawrapper.readthedocs.io/en/latest/index.html

This creates a bar chart using our data. The instructions to create a bar chart are here https://datawrapper.readthedocs.io/en/latest/user-guide/charts/bar-charts.html

In [20]:
chart = dw.BarChart(
    title='AU basketball',
    data=scores_combined_df,
)

# create the chart
chart.create()

Once you've created the chart, you can see it show up in your Datawrapper dashboard. 

**Go to your dashboard on datawrapper.de and click on the new chart you just created. See anything in the data we're missing or want to change?**

--------------

Since there are a couple things in the data we want to change, let's make a new variable called `data_to_chart` that starts as a copy of `scores_combined_df`. Then we can handle the issues individually.

In [48]:
data_to_chart = scores_combined_df

In [52]:
data_to_chart = data_to_chart.reset_index()

data_to_chart = data_to_chart[["opp_name_cleaned", "points_diff"]]

data_to_chart = data_to_chart.sort_values("points_diff",ascending=False)

`data_to_chart.reset_index()` => This fixes the issue that prevented the opponent name column from showing up. The problem was that Pandas was treating that column as an "index" for the dataframe, instead of a column itself. That's because we grouped on it earlier. Good news is it's easy to fix with `.reset_index()`!

`data_to_chart[["opp_name_cleaned", "points_diff"]]` Makes it so we only have the Opponent and Net points columns in our dataframe

`data_to_chart.sort_values("points_diff",ascending=False)` sorts the dataframe by the Net points in descending order


-----------

Once we've made those changes, let's preview the `data_to_chart` dataframe

In [53]:
data_to_chart.head(5)

opp_name_cleaned  points_diff
0       Holy Cross          137
1  Loyola Maryland           70
2             NAVY           70
3    New Hampshire           36
4               CU           35

Now we can update our chart using that cleaned up dataframe. Notice we first make a change to the `chart` variable, then use the `update()` function to push that to the web.

In [54]:
chart.data = data_to_chart
chart.update()

BarChart(chart_type='d3-bars', data=      opp_name_cleaned  points_diff
0           Holy Cross          137
1      Loyola Maryland           70
2                 NAVY           70
3        New Hampshire           36
4                   CU           35
5                  VMI           31
6            LAFAYETTE           30
7                 UMBC           23
8                 UMES           23
9   Saint Francis (PA)           16
10   George Washington           13
11               Brown           12
12              Howard           10
13                 SFU            9
14              ALBANY            8
15   Boston University            5
16                Army           -6
17               Siena           -7
18      William & Mary           -9
19             Hampton          -18
20    Mount St. Mary's          -23
21            Columbia          -23
22           Lafayette          -25
23              LEHIGH          -26
24       Saint Francis          -32
25            BUCKNELL      

**Go to your dashboard and look at the chart again.**


## 🖍 Modifying the chart 

From the library docs, we can see  how to update an existing chart: Updating an existing chart https://datawrapper.readthedocs.io/en/latest/user-guide/chart-operations.html#updating-an-existing-chart


In [31]:
chart.title = "AU basketball has found most success against Holy Cross; worst vs. Lehigh"
chart.update()

BarChart(chart_type='d3-bars', data=              Opponent  Points for  Points against  Net points
0               ALBANY         171             163           8
1                 Army        2055            2061          -6
2             BUCKNELL         275             315         -40
3            Boston U.         649             721         -72
4    Boston University         652             647           5
5                Brown         213             201          12
6             Bucknell        1774            1897        -123
7                   CU         211             176          35
8              Colgate        1594            1666         -72
9             Columbia         206             229         -23
10        George Mason         291             346         -55
11   George Washington         284             271          13
12          Georgetown         375             471         -96
13             Hampton         187             205         -18
14          Holy Cr

----------

After you've done that, you should be able to see the title has updated in the chart in your browser. You don't even need to refresh the page!

---------------

Now we want to update the text on the graphic. We can look to the example on the documentation on how to do this:

The options for a bar chart https://datawrapper.readthedocs.io/en/latest/user-guide/charts/bar-charts.html

🧑‍💻 **In the cell below, complete the following steps:**

Use the example above as a template for updating the following changes (you'll need to look at the options for a bar chart link to see what synatx to use):

    * Introduction: "Cumulative points for and against mens basketball opponents, 2010-2025"
    * Source name: "American University"
    * Source URL: "https://aueagles.com/sports/mens-basketball/stats"
    * Byline: Your name
    * Notes: "Only includes opponents played at least three times. Data not available for 2015 season."

In [38]:
chart.intro = "Cumulative points for and against mens basketball opponents, 2010-2025"
chart.source_name = "American University"
chart.source_url = "https://aueagles.com/sports/mens-basketball/stats"
chart.byline = "YOUR NAME"
chart.notes = "Only includes opponents played at least three times. Data not available for 2015 season."


chart.update()

BarChart(chart_type='d3-bars', data=              Opponent  Net points
14          Holy Cross         137
20     Loyola Maryland          70
22                NAVY          70
24       New Hampshire          36
7                   CU          35
33                 VMI          31
16           LAFAYETTE          30
31                UMBC          23
32                UMES          23
27  Saint Francis (PA)          16
11   George Washington          13
5                Brown          12
15              Howard          10
25                 SFU           9
0               ALBANY           8
4    Boston University           5
1                 Army          -6
29               Siena          -7
35      William & Mary          -9
13             Hampton         -18
21    Mount St. Mary's         -23
9             Columbia         -23
18           Lafayette         -25
17              LEHIGH         -26
26       Saint Francis         -32
2             BUCKNELL         -40
30         Stony Br

-----------------------
Check to make sure this all updated in your browser - if it looks good, congratulations! 

Now we can publish the chart so it's visible to the world!

In [55]:
chart.publish()

BarChart(chart_type='d3-bars', data=      opp_name_cleaned  points_diff
0           Holy Cross          137
1      Loyola Maryland           70
2                 NAVY           70
3        New Hampshire           36
4                   CU           35
5                  VMI           31
6            LAFAYETTE           30
7                 UMBC           23
8                 UMES           23
9   Saint Francis (PA)           16
10   George Washington           13
11               Brown           12
12              Howard           10
13                 SFU            9
14              ALBANY            8
15   Boston University            5
16                Army           -6
17               Siena           -7
18      William & Mary           -9
19             Hampton          -18
20    Mount St. Mary's          -23
21            Columbia          -23
22           Lafayette          -25
23              LEHIGH          -26
24       Saint Francis          -32
25            BUCKNELL      

And we can even see it right here in the notebook now

In [57]:
chart.display()

--------------

# 🎆🎺🎉

Congratulations! You've gone from web scraping to cleaning to analysis to visualization, all automatically!

If there's time, are there any other changes to the chart we should make together?